src/snowflake/cli/api/utils/definition_rendering.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+import os
+from collections import deque
+from dataclasses import dataclass
+from typing import Optional
+
+import yaml
+from jinja2 import Environment, UndefinedError, nodes
+from snowflake.cli.api.utils.graph import Graph
+from snowflake.cli.api.utils.rendering import get_snowflake_cli_jinja_env
+
+
+class Variable:
+    def __init__(self, vars_chain):
+        self._vars_chain = list(vars_chain)
+        self.value: str | None = None
+
+    def get_vars_hierarchy(self) -> deque[str]:
+        return deque(self._vars_chain)
+
+    def get_key(self):
+        return ".".join(self._vars_chain)
+
+    def __hash__(self):
+        return hash(self.get_key())
+
+    def __eq__(self, other):
+        return self.get_key() == other.get_key()
+
+
+@dataclass(eq=False)
+class TemplateGraphNode(Graph.Node):
+    templated_value: str | None = None
+    rendered_value: str | None = None
+    variable: Variable | None = None
+
+    def store_in_context(self, context):
```

Comment:
this whole logic feels a bit complex for what it is. If variable `ctx.env.foo` needs to be stored, then we just need loop through all the names and getOrCreate a dictionary for each (python has that primitive in dict, it's called `setdefault`), and insert the value on the last step. Should be a simple loop. Am I missing anything?

Reply:
can also just make them all defaultdicts: https://stackoverflow.com/a/8702435

Reply:
updated - should be cleaner now